### PR TITLE
Fix more subtle durability issues 

### PR DESCRIPTION
**Pull request by svanoort**
Addresses the following root causes of problems in Pipeline persistence/durability:

* Pipelines which modify the heads in an execution *without* ensuring the corresponding FlowNodes are also saved to disk - this will cause issues when trying to load the execution -- applies only to the fully durable modes (since non-durable just won't persist period)
* Pipelines which modify the `CPSThread.head` `FlowHead` and its `FlowNode` without ensuring the CpsFlowExecution is saved (via `saveOwner()` that saves the `WorkflowRun`).  This causes an NPE loading the program, because the FlowHead may map to a head in the CpsFlowExecution that will not exist at time of deserialization. 
* Provides some documentation on the persistence model so others do not endure my pain
* Removes explicit persistence calls after `FlowHead#setNewHead(FlowNode)` calls because `setNewHead` now forces flownodes to be saved *before* the head is mutated, guaranteeing an always-usable state.
* Expand the fuzzTesting cases
* Consolidate all shutdown-time persistence into the main `CpsFlowExecution#suspendAll()` terminator - *may* help avoid the occasional ConcurrentModificationException we were seeing at serialize, and is more "correct" and safer generally

TARGETTED AT:

* [JENKINS-46986](https://issues.jenkins-ci.org/browse/JENKINS-46986) specifically, and I no longer see NPEs thanks to this

TODO:

- [x] Fix this error seen at shutdown-timer persistence in live testing: 

> Apr 18, 2018 1:01:08 AM hudson.model.Executor finish1
SEVERE: Executor threw an exception
java.lang.IllegalStateException: Jenkins has not been started, or was already shut down
    at jenkins.model.Jenkins.getActiveInstance(Jenkins.java:762)
    at com.cloudbees.hudson.plugins.folder.AbstractFolder.getDescriptor(AbstractFolder.java:605)
    at jenkins.branch.MultiBranchProject.getDescriptor(MultiBranchProject.java:863)
    at jenkins.branch.MultiBranchProject.getDescriptor(MultiBranchProject.java:124)
    at com.cloudbees.hudson.plugins.folder.AbstractFolder.childNameGenerator(AbstractFolder.java:597)
    at com.cloudbees.hudson.plugins.folder.AbstractFolder.getRootDirFor(AbstractFolder.java:640)
    at jenkins.branch.MultiBranchProject.getRootDirFor(MultiBranchProject.java:843)
    at jenkins.branch.MultiBranchProject.getRootDirFor(MultiBranchProject.java:124)
    at hudson.model.AbstractItem.getRootDir(AbstractItem.java:192)
    at hudson.model.Job.getBuildDir(Job.java:845)
    at hudson.model.Run.getRootDir(Run.java:1028)
    at org.jenkinsci.plugins.workflow.job.WorkflowRun.setResult(WorkflowRun.java:738)
    at org.jenkinsci.plugins.workflow.job.WorkflowRun.finish(WorkflowRun.java:747)
    at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:325)
    at hudson.model.ResourceController.execute(ResourceController.java:97)
    at hudson.model.Executor.run(Executor.java:421)

> Apr 18, 2018 1:01:08 AM jenkins.plugins.git.AbstractGitSCMSource getCacheDir
SEVERE: Jenkins instance is null in AbstractGitSCMSource.getCacheDir
Apr 18, 2018 1:01:08 AM jenkins.util.AtmostOneTaskExecutor$1 call
WARNING: null
java.lang.NullPointerException
    at hudson.model.Queue.maintain(Queue.java:1426)
    at hudson.model.Queue$1.call(Queue.java:320)
    at hudson.model.Queue$1.call(Queue.java:317)
    at jenkins.util.AtmostOneTaskExecutor$1.call(AtmostOneTaskExecutor.java:108)
    at jenkins.util.AtmostOneTaskExecutor$1.call(AtmostOneTaskExecutor.java:98)
    at jenkins.security.ImpersonatingExecutorService$2.call(ImpersonatingExecutorService.java:71)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at hudson.remoting.AtmostOneThreadExecutor$Worker.run(AtmostOneThreadExecutor.java:110)
    at java.lang.Thread.run(Thread.java:748)

### CODE REVIEW

**Comment by jglick** [comment]
The NPE from `Queue.maintain` should have been addressed by https://github.com/jenkinsci/jenkins/pull/3365 I think. Just set your core dep higher.

File: src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java [hunk 1]
```diff
@@ -904,6 +910,63 @@ public void evaluate() throws Throwable {
...
+                    assert ((CpsFlowExecution)(run.getExecution())).done;
+                }
+            }
+        });
+
+    }
+
+    /** Test interrupting build by randomly restarting *cleanly* at unpredictable times and verify we stick pick up and resume. */
+    @Test
+    @Ignore //Too long to run as part of main suite
```
[the remark this answers is not in the document]
Comment (svanoort) [reply src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java]: Yeah, you've proposed something special with the Maven build here, I have a pseudo-sample in one of my browser tabs (somewhere between 10 and 10^3 at the moment).  One of those rainy-day projects -- you can always explicitly run via the IDE anyway, or just remove the Ignore annotations

File: doc/persistence.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
+# The Pipeline Persistence Model
```
[the remark this answers is not in the document]
Comment (svanoort) [reply doc/persistence.md]: I know @jglick, pure absurdity, what can I say?

File: pom.xml [hunk 3]
```diff
@@ -141,7 +141,8 @@
         <dependency>
             <groupId>org.jenkins-ci.plugins.workflow</groupId>
             <artifactId>workflow-job</artifactId>
-            <version>2.20</version>
+            <!-- Snapshot for better handling of onLoad errors -->
+            <version>2.21-20180427.223321-5</version>
```
Comment (jglick) [inline pom.xml]: [Pro tip](https://github.com/jenkinsci/pom/blob/master/incrementals.md#usage-in-plugin-poms)

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsBodyExecution.java [hunk 4]
```diff
@@ -367,7 +365,6 @@ public Next receive(Object o) {
             for (BodyExecutionCallback c : callbacks) {
                 c.onSuccess(sc, o);
             }
-
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsBodyExecution.java]: could be reverted

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 5]
```diff
@@ -609,95 +610,89 @@ private synchronized String getHeadsAsString() {
         } else if (myHeads.size() == 0) {
             return "empty-heads";
         } else {
-            return myHeads.entrySet().stream().map(h->h.getKey()+"::"+h.getValue()).collect(Collectors.joining(","));
+            return myHeads.entrySet().stream().map(h -> h.getKey() + "::" + h.getValue()).collect(Collectors.joining(","));
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: unrelated

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 6]
```diff
@@ -609,95 +610,89 @@ private synchronized String getHeadsAsString() {
...
+     */
+    @GuardedBy("this")
+    void createPlaceholderNodes(Throwable failureReason) throws Exception {
+        this.done = true;
+
+        if (this.owner != null) {
+            Queue.Executable ex = owner.getExecutable();
+            if (ex instanceof Run) {
+                Result res = ((Run)ex).getResult();
+                setResult(res != null ? res : Result.FAILURE);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: This seems weird. Normally `null` is a synonym for `SUCCESS`. Why would this not be unconditionally `FAILURE`?

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 7]
```diff
@@ -609,95 +610,89 @@ private synchronized String getHeadsAsString() {
...
-                    }
-                }
+        storage = createStorage();
+        heads = new TreeMap<Integer,FlowHead>();
+        for (Map.Entry<Integer,String> entry : headsSerial.entrySet()) {
+            FlowHead h = new FlowHead(this, entry.getKey());
+
+            FlowNode n = storage.getNode(entry.getValue());
+            if (n != null) {
+                h.setForDeserialize(storage.getNode(entry.getValue()));
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: A bunch of reindentation…not reading.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 8]
```diff
@@ -716,26 +711,45 @@ public boolean canResume() {
...
+                if (canResume()) {
+                    loadProgramAsync(getProgramDataFile());
+                } else {
+                    // TODO if possible, consider trying to close out unterminated blocks to keep existing graph history
+                    // That way we can visualize the graph in some error cases.
+                    LOGGER.log(Level.WARNING, "Pipeline state not properly persisted, cannot resume "+owner.getUrl());
+                    throw new IOException("Cannot resume build -- was not cleanly saved when Jenkins shut down.");
+                }
             }
+        } catch (Exception e) {  // Multicatch ensures that failure to load does not nuke the master
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: BTW this is not technically a multicatch.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 9]
```diff
@@ -906,15 +920,19 @@ public void onFailure(Throwable t) {
...
         if (programPromise == null || !programPromise.isDone()) {
+            // Can't restart cleanly while trying to set up the build
             return true;
         }
         CpsThreadGroup g;
         try {
             g = programPromise.get();
         } catch (Exception x) {
-            return true;
+            // FIXME check this won't cause issues due to depickling delays etc
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: best use `TODO` for consistency

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 10]
```diff
@@ -1207,9 +1228,10 @@ synchronized void onProgramEnd(Outcome outcome) {
         // shrink everything into a single new head
         try {
             if (heads != null) {
+                // Below does not look correct to me
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: Elaborate? Is this a `TODO` comment?

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 11]
```diff
@@ -1221,15 +1243,18 @@ synchronized void onProgramEnd(Outcome outcome) {
...
+            LOGGER.log(Level.WARNING, "Error trying to end execution "+this, ex);
         }
 
         try {
             this.getStorage().flush();
         } catch (IOException ioe) {
             LOGGER.log(Level.WARNING, "Error flushing FlowNodeStorage to disk at end of run", ioe);
         }
+
+        this.persistedClean = Boolean.TRUE;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: or autobox

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 12]
```diff
@@ -1450,6 +1475,13 @@ public String getNextScriptName(String path) {
         return shell.generateScriptName().replaceFirst("[.]groovy$", "");
     }
 
+    /** Has the execution been marked done - note that legacy builds may not have that flag persisted, in which case
+     *  we look for a single FlowEndNode head (see: {@link #isComplete()} and {@link FlowExecution#isComplete()})
+     */
+    public boolean isDoneFlagSet() {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: Not simply `isDone()`?

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 13]
```diff
@@ -1798,15 +1870,32 @@ public void autopersist(@Nonnull FlowNode n) throws IOException {
...
+        if (this.getDurabilityHint().isPersistWithEveryStep()) {
+            saveOwner();
+        }
+    }
+
     /** Save the owner that holds this execution. */
     void saveOwner() {
         try {
             if (this.owner != null && this.owner.getExecutable() instanceof Saveable) {  // Null-check covers some anomalous cases we've seen
                 Saveable saveable = (Saveable)(this.owner.getExecutable());
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: could introduce variable

File: src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java [hunk 14]
```diff
@@ -577,6 +577,9 @@ private void invokeBody(CpsThread cur) {
             // the first one can reuse the current thread, but other ones need to create new heads
             // we want to do this first before starting body so that the order of heads preserve
             // natural ordering.
+
+            // FIXME give this javadocs worth a darn, because this is how we create parallel branches and the docs are crypic.
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java]: `TODO` preferably

File: src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java [hunk 15]
```diff
@@ -790,7 +792,17 @@ public void evaluate() throws Throwable {
         });
     }
 
-    /** Launches the job used for fuzzing in {@link #fuzzTimingDurable()} and {@link #fuzzTimingNonDurable()} -- including the timeout. */
+    private static void assertBuildNotHung(@Nonnull RestartableJenkinsRule story, @Nonnull  WorkflowRun run, int timeOutMillis) throws Exception {
+        if (run.isBuilding()) {
+            try {
+                story.j.waitUntilNoActivityUpTo(timeOutMillis);
+            } catch (AssertionError ase) {  // Allows attaching a debugger here
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java]: Delete now. Debuggers allow you to set exception breakpoints anyway.

File: src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java [hunk 16]
```diff
@@ -865,19 +877,17 @@ public void evaluate() throws Throwable {
             @Override
             public void evaluate() throws Throwable {
                 WorkflowRun run = story.j.jenkins.getItemByFullName(jobName, WorkflowJob.class).getLastBuild();
+                if (run == null) {
+                    return;
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java]: Is this expected?

File: src/test/java/org/jenkinsci/plugins/workflow/cps/PersistenceProblemsTest.java [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+        WorkflowJob job = j.jenkins.createProject(WorkflowJob.class, jobName);
+        job.setDefinition(new CpsFlowDefinition("input 'pause'", true));
+        job.addProperty(new DurabilityHintJobProperty(durabilityHint));
+
+        WorkflowRun run = job.scheduleBuild2(0).getStartCondition().get();
+        ListenableFuture<FlowExecution> listener = run.getExecutionPromise();
+        FlowExecution exec = listener.get();
+        while(exec.getCurrentHeads().isEmpty() || (exec.getCurrentHeads().get(0) instanceof FlowStartNode)) {  // Wait until input step starts
+            System.out.println("Waiting for input step to begin");
+            Thread.sleep(50);
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/PersistenceProblemsTest.java]: Hmm. `SemaphoreStep` is generally preferable for these kinds of tests. Is there a good reason to use `input` instead?

**Comment by svanoort** [comment]
@abayer I can understand wanting the squash but for this on I actually *do* want to show the history, because the commit messages were written to help explain why particular things are being done.

File: src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java (hunk 16 above)
Comment (svanoort) [reply src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java]: Happened in testing -- if the fuzzer kills things *too* early the Run may not even have had time to be created yet (rare but happens).

File: src/test/java/org/jenkinsci/plugins/workflow/cps/PersistenceProblemsTest.java (hunk 17 above)
Comment (svanoort) [reply src/test/java/org/jenkinsci/plugins/workflow/cps/PersistenceProblemsTest.java]: @jglick Yes, Semaphore steps block the CPS thread - input steps follow a more normal pattern.  The durability tests use sleep + semaphore for similar reasons (though sleep should be replaced with input to reduce test times in the near future).

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java (hunk 8 above)
Comment (svanoort) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: Yeah, I need to update that comment since it's just a generic Exception now, good point.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java (hunk 12 above)
Comment (svanoort) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: Nope, we have an isComplete and this is only checking for the flag itself (which may not be 100% trustworthy on its own)

File: pom.xml (hunk 3 above)
Comment (svanoort) [reply pom.xml]: @jglick Since you only added that a few weeks ago and thus it's not "obvious" maybe you meant something [besides 'pro tip'](https://www.urbandictionary.com/define.php?term=Protip)?

**Comment by svanoort** [comment]
CI failure is environmental from the looks of it

File: src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java (hunk 16 above)
Comment (jglick) [reply src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java]: Probably deserves a comment to that effect.

File: src/test/java/org/jenkinsci/plugins/workflow/cps/PersistenceProblemsTest.java (hunk 17 above)
Comment (jglick) [reply src/test/java/org/jenkinsci/plugins/workflow/cps/PersistenceProblemsTest.java]: No, `SemaphoreStep` does not block the CPS thread. Semantically it is very similar to `input`, but much easier to manipulate from test code.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java [hunk 18]
```diff
@@ -729,10 +730,10 @@ public void onLoad(FlowExecutionOwner owner) throws IOException {
         try {
             if (isComplete()) {
                 if (done == Boolean.TRUE && !super.isComplete()) {
-                    LOGGER.log(Level.WARNING, "Completed flow without FlowEndNode: "+this+" heads:"+getHeadsAsString());
+                    LOGGER.log(Level.INFO, "Completed flow without FlowEndNode: "+this+" heads:"+getHeadsAsString());
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java]: Certainly _sounds_ like it should be a warning to me, but maybe I am missing something?

File: src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java (hunk 16 above)
Comment (jglick) [reply src/test/java/org/jenkinsci/plugins/workflow/cps/FlowDurabilityTest.java]: (resolved)

File: src/test/java/org/jenkinsci/plugins/workflow/cps/PersistenceProblemsTest.java (hunk 17 above)
Comment (svanoort) [reply src/test/java/org/jenkinsci/plugins/workflow/cps/PersistenceProblemsTest.java]: @jglick It definitely blocks *something* and the GraphListeners don't evaluate as they ought to.

**Comment by florent-vial** [comment]
I was affected by this. Installing Pipeline: Groovy 2.52 did not fix my symptoms:

`May 05, 2018 1:57:17 PM jenkins.metrics.api.Metrics$HealthChecker execute
WARNING: Some health checks are reporting as unhealthy: [thread-deadlock : [Running CpsFlowExecution[Owner[MY_PIPELINE_JOB/7521:MY_PIPELINE_JOB #7521]] locked on org.jenkinsci.plugins.workflow.job.WorkflowRun@232fe592 (owned by WorkflowRun.copyLogs [#4] (MY_PIPELINE_JOB #7521)):
         at org.jenkinsci.plugins.workflow.job.WorkflowRun.save(WorkflowRun.java:1245)
         at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.saveOwner(CpsFlowExecution.java:1925)
         at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.saveExecutionIfDurable(CpsFlowExecution.java:1906)
         at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.removeHead(CpsFlowExecution.java:1082)
         at org.jenkinsci.plugins.workflow.cps.FlowHead.remove(FlowHead.java:177)
         at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.subsumeHead(CpsFlowExecution.java:1097)
         at org.jenkinsci.plugins.workflow.cps.CpsStepContext$2.onSuccess(CpsStepContext.java:436)
         at org.jenkinsci.plugins.workflow.cps.CpsStepContext$2.onSuccess(CpsStepContext.java:395)
         at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution$4$1.run(CpsFlowExecution.java:909)
         at org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$1.run(CpsVmExecutorService.java:35)
         at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
         at java.util.concurrent.FutureTask.run(FutureTask.java:266)
         at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:131)
         at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
         at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
         at java.util.concurrent.FutureTask.run(FutureTask.java:266)
         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
         at java.lang.Thread.run(Thread.java:748)
, WorkflowRun.copyLogs [#4] (MY_PIPELINE_JOB #7521) locked on org.jenkinsci.plugins.workflow.cps.CpsFlowExecution@757d1262 (owned by Running CpsFlowExecution[Owner[MY_PIPELINE_JOB/7521:MY_PIPELINE_JOB #7521]]):
         at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution$ConverterImpl.marshal(CpsFlowExecution.java:1603)
         at hudson.util.XStream2$AssociatedConverterImpl.marshal(XStream2.java:461)
         at com.thoughtworks.xstream.core.AbstractReferenceMarshaller.convert(AbstractReferenceMarshaller.java:69)
         at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:58)
         at com.thoughtworks.xstream.core.AbstractReferenceMarshaller$1.convertAnother(AbstractReferenceMarshaller.java:84)
         at hudson.util.RobustReflectionConverter.marshallField(RobustReflectionConverter.java:265)
         at hudson.util.RobustReflectionConverter$2.writeField(RobustReflectionConverter.java:252)
         at hudson.util.RobustReflectionConverter$2.visit(RobustReflectionConverter.java:224)
         at com.thoughtworks.xstream.converters.reflection.PureJavaReflectionProvider.visitSerializableFields(PureJavaReflectionProvider.java:138)
         at hudson.util.RobustReflectionConverter.doMarshal(RobustReflectionConverter.java:209)
         at hudson.util.RobustReflectionConverter.marshal(RobustReflectionConverter.java:150)
         at com.thoughtworks.xstream.core.AbstractReferenceMarshaller.convert(AbstractReferenceMarshaller.java:69)
         at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:58)
         at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:43)
         at com.thoughtworks.xstream.core.TreeMarshaller.start(TreeMarshaller.java:82)
         at com.thoughtworks.xstream.core.AbstractTreeMarshallingStrategy.marshal(AbstractTreeMarshallingStrategy.java:37)
         at com.thoughtworks.xstream.XStream.marshal(XStream.java:1026)
         at com.thoughtworks.xstream.XStream.marshal(XStream.java:1015)
         at com.thoughtworks.xstream.XStream.toXML(XStream.java:988)
         at hudson.XmlFile.write(XmlFile.java:193)
         at org.jenkinsci.plugins.workflow.support.PipelineIOUtils.writeByXStream(PipelineIOUtils.java:30)
         at org.jenkinsci.plugins.workflow.job.WorkflowRun.save(WorkflowRun.java:1256)
         at org.jenkinsci.plugins.workflow.job.WorkflowRun.saveWithoutFailing(WorkflowRun.java:1236)
         at org.jenkinsci.plugins.workflow.job.WorkflowRun.copyLogs(WorkflowRun.java:612)
         at org.jenkinsci.plugins.workflow.job.WorkflowRun.access$500(WorkflowRun.java:144)
         at org.jenkinsci.plugins.workflow.job.WorkflowRun$3.run(WorkflowRun.java:410)
         at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
         at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
         at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
         at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
         at java.lang.Thread.run(Thread.java:748)
`
